src/theory/uf/equality_engine.cpp

```diff
@@ -1024,15 +1036,51 @@ void EqualityEngine::explainPredicate(TNode p, bool polarity,
                     << std::endl;
   // Must have the term
   Assert(hasTerm(p));
+  std::map<std::pair<EqualityNodeId, EqualityNodeId>, EqProof*> cache;
   // Get the explanation
-  getExplanation(getNodeId(p), polarity ? d_trueId : d_falseId, assertions,
-                 eqp);
+  getExplanation(
+      getNodeId(p), polarity ? d_trueId : d_falseId, assertions, cache, eqp);
 }
 
-void EqualityEngine::getExplanation(EqualityNodeId t1Id, EqualityNodeId t2Id,
-                                    std::vector<TNode>& equalities,
-                                    EqProof* eqp) const {
-  Debug("equality") << d_name << "::eq::getExplanation(" << d_nodes[t1Id] << "," << d_nodes[t2Id] << ")" << std::endl;
+void EqualityEngine::getExplanation(
+    EqualityNodeId t1Id,
+    EqualityNodeId t2Id,
+    std::vector<TNode>& equalities,
+    std::map<std::pair<EqualityNodeId, EqualityNodeId>, EqProof*>& cache,
+    EqProof* eqp) const
+{
+  Trace("eq-exp") << d_name << "::eq::getExplanation(" << d_nodes[t1Id] << ","
+                  << d_nodes[t2Id] << ") size = " << cache.size() << std::endl;
+
+  // We order the ids, since explaining t1 = t2 is the same as explaining
+  // t2 = t1.
+  std::pair<EqualityNodeId, EqualityNodeId> cacheKey = std::minmax(t1Id, t2Id);
+  std::map<std::pair<EqualityNodeId, EqualityNodeId>, EqProof*>::iterator it =
+      cache.find(cacheKey);
+  if (it != cache.end())
+  {
+    // copy one level
+    if (eqp)
+    {
+      if (it->second)
+      {
+        eqp->d_node = it->second->d_node;
+        eqp->d_id = it->second->d_id;
+        eqp->d_children.insert(eqp->d_children.end(),
+                               it->second->d_children.begin(),
+                               it->second->d_children.end());
+      }
+      else
+      {
+        // We may have cached null in its place, create the trivial proof now.
+        Assert(d_nodes[t1Id] == d_nodes[t2Id]);
+        Assert(eqp->d_id == MERGED_THROUGH_REFLEXIVITY);
+        eqp->d_node = d_nodes[t1Id];
+      }
+    }
+    return;
+  }
+  cache[cacheKey] = eqp;
```

Comment:
Setting the cache here before filling in the proof seems a little dangerous.  Can you double-check that there's no path where you end up caching an incomplete or invalid proof?  Alternatively - and to make sure we don't make mistakes in future revisions - maybe we should wait until the end of the function to set the cache.

Reply:
I've checked this is the case. At a higher level, the contract of the function says that `eqp` must always be filled in with a proper proof if it is non-null, so whatever case where eqp was incomplete or invalid, it would be breaking the contract of the overall function.